### DISCUSSION TITLE
The computed specified array subscript in react will not update automatically

### DISCUSSION THREAD

**Discussion by StringKe** [mobxjs/mobx#3150] [Q&A] [answered]
```typescript
import { makeAutoObservable } from 'mobx';

import { observer } from 'mobx-react-lite';

class TestData {
	list: any[] = [];
	listIndex = -1;

	constructor() {
		makeAutoObservable(this);
	}

	get activeItem() {
		return this.list.reduce((result, item, index, arr) => {
			if (index === this.listIndex) {
				result = item;
				// End the loop early
				arr.splice(1);
			}
			return result;
		});
	}

	changeIndex() {
		this.listIndex += 1;
	}

	changeListData() {
		this.list[0].list[0] = { a: 1, b: 2 };
	}

	changeList() {
		this.list.push({
			id: Math.floor(Math.random() * 100000),
			data: {},
			list: [{ a: 1 }],
		});
	}
}

const testData = new TestData();

const TestComponent = observer(() => {
	const item = testData.activeItem;
	if (!item) {
		return null;
	}

	return (
		<div>
			{item.id} {item.list}
		</div>
	);
});

```

The computed activeItem will not respond correctly, what do I need to do to achieve this function?

**Comment by urugator** [comment, marked as answer]
Dunno what the intention is, but `computed`  shouldn't mutate observables  (`arr.splice(1)`). Maybe you just wanna do the following?
```javascript
get activeItem() {
  return this.list[this.listIndex];
}
```

**Reply by StringKe** [reply]
I have tried this. It is the same as when I use reduce. When changeListData updates the data, the component will not re-render.

My function is similar to selecting data in an array and using subscripts to record the currently selected data.

**Reply by kubk** [reply]
@StringKe Please try to reproduce it here: https://codesandbox.io/s/minimal-mobx-react-project-ppgml?file=/index.js

**Reply by urugator** [reply]
Btw `changeListData` in your example mutates a nested list (`this.list[0].list[0]`), nothing (neither the computed or the component) depends on this nested array, therefore nothing will re-compute/re-render.

**Reply by StringKe** [reply]
https://codesandbox.io/s/elastic-browser-rk0vm?file=/index.js

The component will not refresh when the "update element" button is clicked

**Reply by urugator** [reply]
You're missing `observer` on `PageRender`

**Reply by StringKe** [reply]
PageRender is a third-party package, it does not use mobx

**Reply by urugator** [reply]
Then you can't pass observables to that component, you can use `<PageRender page={toJS(selectPage)} />`, please see:
https://mobx.js.org/react-integration.html#dont-pass-observables-into-components-that-arent-observer
https://github.com/mobxjs/mobx-react/issues/484#issuecomment-391685354

**Reply by StringKe** [reply]
okay, thank you,
